docs/04-reference/winglang-spec.md

```diff
@@ -869,30 +869,223 @@ let f = (arg1: num, var arg2: num) => {
+
+```js
+let increment = (x: num, by: num = 1): num {
+  return x + by;
+}
+```
+
+Non-optional arguments can only be used before all optional arguments:
+
+```js
+let my_fun = (a: str, x?: num, y: str): void = { /* ... */ };
+//-----------------------------^^^^^^ ERROR: cannot declare a non-optional argument after an optional
+```
+
+If a function uses a keyword argument struct as the last argument, and there are other optional arguments before,
+it also has to be declared as optional:
+
+```js
+let parse_int = (x: str, radix: num?, opts?: ParseOpts): num { /* ... */ };
+// or
+let parse_int = (x: str, radix: num = 10, opts: ParseOpts = ParseOpts {}): num { /* ... */ };
+```
+
+##### 1.7.1.5 Declaration in function return types
+
+If a function returns an optional type, use the `return nil;` statement to indicate that the value is
+not defined.
+
+> This is currently the only scenario in the language where `nil` is allowed.
+
+```js
+struct Name { first: str, last: str };
+
+let try_parse_name = (full_name: str): Name? => {
+  let parts = full_name.split(" ");
+  if parts.len < 2 {
+    return nil;
+  }
+
+  return Name { first: parts.at(0), last: parts.at(1) };
+}
+```
+
+#### 1.7.2 Testing using `x?`
+
+Object of type `str?` cannot be used within `==` or `!=` expressions because we don't have a null
+value. Instead, use the the `?` operator to get a `bool` that indicates if the optional is defined
+or not:
+
+```js
+let is_address_defined = my_person.address?; // type is `bool`
+
+// or within a condition
+if my_person.address? {
+  print("address is defined but i do not care what it is");
+}
+
+// can be negated
+if !my_person.address? {
+  print("address is not defined");
+}
+```
+
+#### 1.7.3 Unwrapping using `if let` and `let else`
+
+The `if let` statement can be used to test if an optional is defined an *unwrap* it into a
+non-optional variable defined inside the block:
+
+```js
+if let address = my_person.address {
+  print(address.len);
+  print(address); // address is type `str`
+}
+```
+
+> NOTE: `if let` is not the same as `if`. For example, we currently don't support specifying multiple conditions, or unwrapping multiple optionals. This is something we might consider in the future.
+
+The `let .. else` statement can be used to conditionally execute a block if an optional
+does not have a value:
+
+```js
+let address = my_person.address else {
+  throw("address is required");
+};
+
+print("address is ${address}");
+```
+
+#### 1.7.4 Unwrapping or default value using `??`
+
+The `??` operator can be used to unwrap or provide a default value. This returns a value of `T` that can safely be used.
+
+```js
+let address: str = my_person.address ?? "Planet Earth";
+```
+
+#### 1.7.5 Optional chaining using `?.`
+
+The `?.` syntax can be used for optional chaining. Optional chaining returns a value of type `T?` which must be unwrapped in order to be used.
+
+```js
+if let ip_address = options.networking?.ip_address {
```

Comment:
To be clear about what's happening here (correct me where I'm wrong):
`networking` is of type `T1?`
`ip_address` is of type `T2`
The expression `options.networking` is a `T1?`
The expression `options.networking?` is a `T1`
The expression `options.networking?.ip_address` is a `T2?`
so `if let ip_address = xxx` is unwrapping a `T2?` into `ip_address`.
So if I get this correctly given a `T1?.T2?.T3.T4` we get a `T4?`.

Also it's a bit confusing that `?` is used both for optional chaining and for evaluating to `bool`:
<code>T1?.T2; // This is a `T2?`</code>
<code>T1?; // This is a `bool`</code>
<code>T1?.T2?; // This is a `bool`</code>
I guess we can try this a bit and see how it feels...